File: pufa-backend/scripts/generar-informe-word.js
const fs = require('fs');
const path = require('path');
const { Document, Packer, Paragraph, HeadingLevel, TextRun, ImageRun } = require('docx');

const reportDir = path.join(__dirname, '..', 'reportes');
const outputPath = path.join(reportDir, 'seguimiento_practica_pufa.docx');

if (!fs.existsSync(reportDir)) {
  fs.mkdirSync(reportDir, { recursive: true });
}

const semanas = [
  'Semana del 16 al 20 de marzo: Se dio inicio al desarrollo del backend. Se estructuro la base del proyecto en NestJS, la configuracion inicial de TypeORM y las primeras validaciones de conexion a base de datos.',
  'Semana del 23 al 27 de marzo: Se implementaron modulos base de autenticacion y registro. Se trabajo sobre DTOs, controladores y validacion de datos para el flujo de usuarios.',
  'Semana del 30 de marzo al 3 de abril: Se avanzo con entidades y relaciones principales (proyectos y tramites), ajustando el modelo de datos para soportar la creacion y vinculacion correcta.',
  'Semana del 6 al 10 de abril: Se desarrollo la gestion de documentos: carga, listado y descarga, incluyendo almacenamiento en servidor y hash de integridad.',
  'Semana del 13 al 17 de abril: Se agrego la generacion del recibo en PDF para tramites, con endpoint de descarga para uso desde la interfaz.',
  'Semana del 20 al 24 de abril: Se integro frontend con backend en formularios criticos, corrigiendo payloads y mapeos de campos para registro y creacion de tramites.',
  'Semana del 27 de abril al 1 de mayo: Se fortalecio el modulo administrativo para la gestion de imagenes de locaciones (subir, listar, eliminar, reordenar) y se corrigieron errores de modal.',
  'Semana del 4 al 8 de mayo: Se completaron mejoras de portafolio proveedor (servicios y galeria), visualizacion de tramites en detalle de proyecto y ajuste del script de arranque productivo.',
];

const capturas = [
  'Login e ingreso: pantalla de inicio de sesion centrada (public/iniciar-sesion/index.html).',
  'Registro: formulario completo y mensaje de confirmacion de alta (public/registro/index.html).',
  'Crear tramite: seleccion de locacion existente y seccion de adjuntos (public/crear-tramite/index.html).',
  'Detalle de tramite: listado de documentos y boton de descarga de recibo PDF (public/detalle-tramite/index.html).',
  'Detalle de proyecto: seccion de tramites asociados al proyecto (public/detalle-proyecto/index.html).',
  'Crear proyecto: selector de servicio del proveedor y carga de documentos (public/crear-proyecto/index.html).',
  'Administracion: modal de imagenes de locacion con carga y gestion (public/admin/admin-4.html).',
  'Proveedor/Portafolio: bloques de servicios y galeria con acciones de agregar (public/proveedor/proveedor-2.html).',
];

const imagenesFront = [
  {
    file: '1.jpg',
    caption: 'Imagen 14. Portada visual para la pagina de inicio del portal.',
  },
  {
    file: '2.jpg',
    caption: 'Imagen 15. Imagen de apoyo para la seccion institucional del frontend.',
  },
  {
    file: '3.jpeg',
    caption: 'Imagen 16. Imagen de apoyo para formularios de registro y presentacion del sistema.',
  },
  {
    file: '4.jpg',
    caption: 'Imagen 17. Imagen panoramica para secciones de proyectos y tramites.',
  },
  {
    file: '5.jpg',
    caption: 'Imagen 18. Imagen de apoyo para el portal de contenidos y galeria visual.',
  },
];

function imageParagraph(fileName) {
  const imagePath = path.join(__dirname, '..', 'imgs', fileName);
  const imageBuffer = fs.readFileSync(imagePath);

  return new Paragraph({
    children: [
      new ImageRun({
        data: imageBuffer,
        transformation: { width: 600, height: 338 },
      }),
    ],
  });
}

const children = [
  new Paragraph({ text: 'Seguimiento de Trabajo - Proyecto PUFA', heading: HeadingLevel.TITLE }),
  new Paragraph({ text: 'Fecha de elaboracion: 07 de mayo de 2026' }),
  new Paragraph({ text: '' }),
  new Paragraph({ text: '1. Seguimiento semanal', heading: HeadingLevel.HEADING_1 }),
  ...semanas.flatMap((s) => [
    new Paragraph({
      children: [new TextRun({ text: s })],
      bullet: { level: 0 },
    }),
    new Paragraph({ text: '' }),
  ]),
  new Paragraph({ text: '2. Evidencias visuales del frontend', heading: HeadingLevel.HEADING_1 }),
  ...imagenesFront.flatMap((imagen) => [
    imageParagraph(imagen.file),
    new Paragraph({ text: imagen.caption, alignment: 'center' }),
    new Paragraph({ text: '' }),
  ]),
  new Paragraph({ text: '3. Capturas recomendadas por apartado', heading: HeadingLevel.HEADING_1 }),
  ...capturas.map((c, idx) => new Paragraph({ text: `${idx + 1}. ${c}` })),
  new Paragraph({ text: '' }),
  new Paragraph({ text: '4. Evidencias tecnicas sugeridas', heading: HeadingLevel.HEADING_1 }),
  new Paragraph({ text: 'Complementar con capturas de endpoints y archivos clave en backend: src/modules/documentos/documentos.controller.ts, src/modules/tramites/tramites.controller.ts, src/app.controller.ts.' }),
];

const doc = new Document({
  sections: [{ children }],
});

Packer.toBuffer(doc)
  .then((buffer) => {
    fs.writeFileSync(outputPath, buffer);
    console.log(`OK: ${outputPath}`);
  })
  .catch((error) => {
    console.error('ERROR:', error.message);
    process.exit(1);
  });

File: pufa-backend/scripts/generar_informe_word.py
from docx import Document
from docx.shared import Pt
from datetime import date
from pathlib import Path


def add_heading(doc, text, level=1):
    h = doc.add_heading(text, level=level)
    if h.runs:
        h.runs[0].font.name = "Calibri"
    return h


def add_paragraph(doc, text, bold=False):
    p = doc.add_paragraph()
    run = p.add_run(text)
    run.bold = bold
    run.font.name = "Calibri"
    run.font.size = Pt(11)
    return p


def add_bullet(doc, text):
    p = doc.add_paragraph(text, style="List Bullet")
    for run in p.runs:
        run.font.name = "Calibri"
        run.font.size = Pt(11)
    return p


def main():
    doc = Document()

    add_heading(doc, "Informe de seguimiento de desarrollo - Proyecto Alcaldía", level=0)
    add_paragraph(doc, f"Fecha de elaboración: {date.today().strftime('%d/%m/%Y')}")

    add_heading(doc, "1. Seguimiento semanal", level=1)

    add_heading(doc, "Semana del 16 al 20 de marzo", level=2)
    add_paragraph(doc, "Se dio inicio al desarrollo del backend.")
    add_bullet(doc, "Estructura inicial de NestJS y configuración base del proyecto.")
    add_bullet(doc, "Configuración de TypeORM y conexión a PostgreSQL.")
    add_bullet(doc, "Verificación temprana de conectividad de base de datos.")

    add_heading(doc, "Semana del 23 al 27 de marzo", level=2)
    add_paragraph(doc, "Se implementó la capa inicial de autenticación y usuarios.")
    add_bullet(doc, "Endpoints de autenticación y registro.")
    add_bullet(doc, "DTOs y validaciones de entrada.")
    add_bullet(doc, "Guardias JWT para protección de rutas.")

    add_heading(doc, "Semana del 30 de marzo al 3 de abril", level=2)
    add_paragraph(doc, "Se consolidaron entidades y relaciones del dominio principal.")
    add_bullet(doc, "Modelado de proyectos y trámites.")
    add_bullet(doc, "Ajustes de relaciones TypeORM para evitar errores de propiedades.")
    add_bullet(doc, "Persistencia de vínculo trámite-proyecto (proyecto_id).")

    add_heading(doc, "Semana del 6 al 10 de abril", level=2)
    add_paragraph(doc, "Se habilitó la gestión documental en backend y frontend.")
    add_bullet(doc, "Subida, listado y descarga de documentos.")
    add_bullet(doc, "Almacenamiento de archivos en carpeta uploads.")
    add_bullet(doc, "Integración con formularios de creación de trámite/proyecto.")

    add_heading(doc, "Semana del 13 al 17 de abril", level=2)
    add_paragraph(doc, "Se agregó la generación de recibo bancario en PDF.")
    add_bullet(doc, "Endpoint para generar y descargar recibo PDF por trámite.")
    add_bullet(doc, "Integración de botón de descarga en detalle de trámite.")

    add_heading(doc, "Semana del 20 al 24 de abril", level=2)
    add_paragraph(doc, "Se realizaron ajustes funcionales de frontend y payloads.")
    add_bullet(doc, "Corrección de campos enviados desde registro de usuarios.")
    add_bullet(doc, "Alineación de nombres de campos frontend-backend.")
    add_bullet(doc, "Mejoras de navegación entre crear trámite y detalle de proyecto.")

    add_heading(doc, "Semana del 27 de abril al 1 de mayo", level=2)
    add_paragraph(doc, "Se fortaleció el módulo administrativo de locaciones e imágenes.")
    add_bullet(doc, "Correcciones de modal de imágenes y referencias DOM.")
    add_bullet(doc, "Carga de imágenes para locaciones desde panel admin.")
    add_bullet(doc, "Operaciones de listado, borrado y orden en galería.")

    add_heading(doc, "Semana del 4 al 8 de mayo", level=2)
    add_paragraph(doc, "Se incorporaron mejoras de portafolio proveedor y despliegue.")
    add_bullet(doc, "Cambio de enfoque de 'Agregar Proyecto' a 'Agregar servicios'.")
    add_bullet(doc, "Soporte de galería del proveedor y selección de servicios en proyectos.")
    add_bullet(doc, "Ajuste del script start:prod para compilar y luego ejecutar.")

    add_heading(doc, "2. Capturas recomendadas por apartado", level=1)

    add_heading(doc, "2.1 Portada / Autenticación", level=2)
    add_bullet(doc, "Pantalla de inicio de sesión centrada (public/iniciar-sesion/index.html).")
    add_bullet(doc, "Versión responsive en resolución móvil.")

    add_heading(doc, "2.2 Registro de usuario", level=2)
    add_bullet(doc, "Formulario completo de registro (public/registro/index.html).")
    add_bullet(doc, "Evidencia de validación o registro exitoso.")

    add_heading(doc, "2.3 Trámites y documentos", level=2)
    add_bullet(doc, "Crear trámite con selección de locación existente (public/crear-tramite/index.html).")
    add_bullet(doc, "Subida de documento en trámite/proyecto.")
    add_bullet(doc, "Detalle de trámite con botón de descarga de PDF (public/detalle-tramite/index.html).")

    add_heading(doc, "2.4 Proyectos", level=2)
    add_bullet(doc, "Detalle de proyecto mostrando trámites asociados (public/detalle-proyecto/index.html).")
    add_bullet(doc, "Crear proyecto con selección de servicio (public/crear-proyecto/index.html).")

    add_heading(doc, "2.5 Administración", level=2)
    add_bullet(doc, "Modal de imágenes en administración de locaciones (public/admin/admin-4.html).")
    add_bullet(doc, "Proceso de subida de imagen desde el modal.")

    add_heading(doc, "2.6 Proveedor / Portafolio", level=2)
    add_bullet(doc, "Listado de servicios y galería del proveedor (public/proveedor/proveedor-2.html).")
    add_bullet(doc, "Formulario para agregar servicio y evidencia de carga de imagen en galería.")

    add_heading(doc, "3. Recomendaciones de presentación", level=1)
    add_bullet(doc, "Nombrar capturas de forma secuencial: 01-login, 02-registro, 03-crear-tramite, etc.")
    add_bullet(doc, "Usar resolución horizontal (1280x720) para páginas completas.")
    add_bullet(doc, "Para modales, tomar captura con zoom suficiente para leer campos y botones.")
    add_bullet(doc, "Incluir una captura por flujo con estado final exitoso (guardado/subida/descarga).")

    output_dir = Path(__file__).resolve().parents[1] / "reportes"
    output_dir.mkdir(parents=True, exist_ok=True)
    output_file = output_dir / "informe_seguimiento_proyecto_alcaldia.docx"
    doc.save(output_file)

    print(f"Documento generado: {output_file}")


if __name__ == "__main__":
    main()
